Sources/Filters/General/ContourTriangulator/helper.js

```diff
@@ -0,0 +1,2290 @@
+  const v = [];
+  let l1;
+  let l2;
+
+  const minVal = [0, 0];
+
+  verts[0] = 0;
+  verts[1] = 0;
+
+  const n = poly.length;
+  points.getPoint(poly[n - 1], p2);
+  points.getPoint(poly[0], p1);
+
+  vtkMath.subtract(p1, p2, v1);
+  l1 = Math.sqrt(vtkMath.dot(v1, v1));
+
+  for (let j = 0; j < n; j++) {
+    let k = j + 1;
+    if (k === n) {
+      k = 0;
+    }
+
+    points.getPoint(poly[k], p2);
+    vtkMath.subtract(p2, p1, v2);
+    l2 = Math.sqrt(vtkMath.dot(v2, v2));
+
+    vtkMath.cross(v1, v2, v);
+    const b = vtkMath.dot(v, normal);
+
+    if (b < 0 && l1 * l2 > 0) {
+      // Dot product is |v1||v2|cos(theta), range [-1, +1]
+      const val = vtkMath.dot(v1, v2) / (l1 * l2);
+      if (val < minVal[0]) {
+        minVal[1] = minVal[0];
+        minVal[0] = val;
+        verts[1] = verts[0];
+        verts[0] = j;
+      }
+    }
+
+    // Rotate to the next point
+    p1[0] = p2[0];
+    p1[1] = p2[1];
+    p1[2] = p2[2];
+    v1[0] = v2[0];
+    v1[1] = v2[1];
+    v1[2] = v2[2];
+    l1 = l2;
+  }
+}
+
+// ---------------------------------------------------
+/**
+ * Find two valid cuts between outerPoly and innerPoly.
+ * Used by vtkCCSCutHoleyPolys.
+ *
+ * @param {Number[]} polys
+ * @param {Number[]} polyGroup
+ * @param {Number} outerPolyId
+ * @param {Number} innerPolyId
+ * @param {vtkPoints} points
+ * @param {Vector3} normal
+ * @param {Number[][]} cuts
+ * @param {Boolean} exhaustive
+ */
+export function vtkCCSFindCuts(
+  polys,
+  polyGroup,
+  outerPolyId,
+  innerPolyId,
+  points,
+  normal,
+  cuts,
+  exhaustive
+) {
+  const outerPoly = polys[outerPolyId];
+  const innerPoly = polys[innerPolyId];
+  const innerSize = innerPoly.length;
+  // Find the two sharpest points on the inner poly
+  // TODO: Check
+  const verts = [];
+  vtkCCSFindSharpestVerts(innerPoly, points, normal, verts);
+
+  // A list of cut locations according to quality
+  // TODO: Check
+  // typedef std::pair<double, size_t> vtkCCSCutLoc;
+  // std::vector<vtkCCSCutLoc> cutlist(outerPoly.size());
+  const cutlist = [];
+  cutlist.length = outerPoly.length;
+
+  // Search for potential cuts (need to find two cuts)
+  let cutId = 0;
+  cuts[0][0] = 0;
+  cuts[0][1] = 0;
+  cuts[1][0] = 0;
+  cuts[1][1] = 0;
+
+  let foundCut = false;
+  for (cutId = 0; cutId < 2; cutId++) {
+    const count = exhaustive ? innerSize : 3;
+
+    for (let i = 0; i < count && !foundCut; i++) {
+      // Semi-randomize the search order
+      // eslint-disable-next-line no-bitwise
+      let j = (i >> 1) + (i & 1) * ((innerSize + 1) >> 1);
+      // Start at the best first point
+      j = (j + verts[cutId]) % innerSize;
+
+      for (let kk = 0; kk < outerPoly.length; kk++) {
+        const q = vtkCCSCutQuality(outerPoly, innerPoly, kk, j, points);
+        // cutlist[kk].first = q;
+        // cutlist[kk].second = kk;
+        cutlist[kk] = [q, kk];
+      }
+
+      // TODO: Check
+      // std::sort(cutlist.begin(), cutlist.end());
+      cutlist.sort();
```

Comment:
use a safe sort ?

Reply:
and make a pass everywhere in your code to make sure there is no other sort

Reply:
What do you mean by safe sort?